Feature/ET-796 RSVP add constant to enable

Luc45 opened the pull request:
This PR enables RSVP to be viewed as Early Access for QA purposes.

Ticket: https://moderntribe.atlassian.net/browse/ET-796

@juanfra, you might want to use these for the new RSVP for now:

- Template: `views/tickets/rsvp-early-access.php`
- CSS: `resources/postcss/rsvp-early-access.css`
- JS: `resources/js/rsvp-early-access.js`

To enable Early Access for RSVPs, drop the mu-plugin in `mu-plugin/rsvp-early-access.php`:

```php
/**
 * Plugin Name: RSVP Early Access
 */
add_action( 'wp', function () {
	tribe( \Tribe\Tickets\RSVP\Early_Access::class )->set_rsvp_early_access( true );
} );
```

_PS: I'm setting the base to `tests/page-objects` for merge here on GitHub to have a better diff of the files changed on this PR, since I've merged `tests/page-objects` in this branch._

tr1b0t commented on src/Tribe/Main.php:
```diff
@@ -371,6 +371,9 @@ public function bind_implementations() {
 		// Views V2
 		tribe_register_provider( Tribe\Tickets\Events\Views\V2\Service_Provider::class );
+
+		// RSVP
```
:no_entry_sign: Inline comments must end in full-stops, exclamation marks, or question marks

<sub>hash: 420b889634ca48c156254152dcb54d9f</sub>

tr1b0t commented on src/Tribe/RSVP/Early_Access.php:
```diff
@@ -0,0 +1,130 @@
+<?php
+
```
:no_entry_sign: Missing file doc comment

<sub>hash: 752a821097eaeace2e5d7c35c6ead528</sub>

tr1b0t commented on src/Tribe/RSVP/Early_Access.php:
```diff
@@ -0,0 +1,130 @@
+	 * @since TBD
+	 *
+	 * @param bool $is_early_access
+	 *
```
:no_entry_sign: Missing parameter comment

<sub>hash: 97e8b1c9a862a20b8922d745f6b16465</sub>

tr1b0t commented on src/Tribe/RSVP/Early_Access.php:
```diff
@@ -0,0 +1,130 @@
+		}
+
+		/* @see \Tribe__Tickets__RSVP::register_resources */
+		wp_deregister_style( 'event-tickets-rsvp' );
```
:warning: This comment is 43% valid code; is this commented out code?

<sub>hash: c978571a6a0d36c1ee5afa89471de786</sub>

tr1b0t commented on src/Tribe/RSVP/Early_Access.php:
```diff
@@ -0,0 +1,130 @@
+		wp_deregister_script( 'event-tickets-rsvp' );
+
+		/* @see \Tribe__Tickets__Assets::enqueue_scripts */
+		wp_deregister_style( 'event-tickets-tickets-rsvp-css' );
```
:warning: This comment is 43% valid code; is this commented out code?

<sub>hash: 596ae41bba853937776b8c62b7f1595f</sub>

tr1b0t commented on src/Tribe/RSVP/Early_Access.php:
```diff
@@ -0,0 +1,130 @@
+		return $file;
+	}
+}
\ No newline at end of file
```
:no_entry_sign: File must end with a newline character

<sub>hash: e5eaae9796773696f921e95b55ee3b6c</sub>

tr1b0t commented on src/Tribe/RSVP/Service_Provider.php:
```diff
@@ -0,0 +1,44 @@
+<?php
+
```
:no_entry_sign: Missing file doc comment

<sub>hash: 3fd9656f09002708c193e5794104dc81</sub>

tr1b0t commented on src/Tribe/RSVP/Service_Provider.php:
```diff
@@ -0,0 +1,44 @@
+	 */
+	private function register_early_access() {
+		add_action( 'init', [
+			$this->container->make( Early_Access::class ),
```
* :no_entry_sign: Opening parenthesis of a multi-line function call must be the last content on the line
* :no_entry_sign: There must be a single space before the last parenthesis
* :no_entry_sign: Only one argument is allowed per line in a multi-line function call

<sub>hash: ddce73dcb8fb40becbc4f37a15d03064</sub>

tr1b0t commented on src/Tribe/RSVP/Service_Provider.php:
```diff
@@ -0,0 +1,44 @@
+		add_action( 'init', [
+			$this->container->make( Early_Access::class ),
+			'maybe_register_early_access_assets'
+		]);
```
* :no_entry_sign: Each array item in a multi-line array declaration must end in a comma
* :no_entry_sign: Comma required after last value in array declaration

<sub>hash: 607f52e1958e8e661a8ed35000f44d60</sub>

tr1b0t commented on src/Tribe/RSVP/Service_Provider.php:
```diff
@@ -0,0 +1,44 @@
+			$this->container->make( Early_Access::class ),
+			'maybe_register_early_access_assets'
+		]);
+
```
:no_entry_sign: Closing parenthesis of a multi-line function call must be on a line by itself

<sub>hash: ed769dfe1733e1433a726f88a08fd6c6</sub>

tr1b0t commented on src/Tribe/RSVP/Service_Provider.php:
```diff
@@ -0,0 +1,44 @@
+		]);
+
+		add_action( 'wp_enqueue_scripts', [
+			$this->container->make( Early_Access::class ),
```
* :no_entry_sign: Opening parenthesis of a multi-line function call must be the last content on the line
* :no_entry_sign: Only one argument is allowed per line in a multi-line function call

<sub>hash: 63388b0e628d7a5d2ef192256c124b0e</sub>

tr1b0t commented on src/Tribe/RSVP/Service_Provider.php:
```diff
@@ -0,0 +1,44 @@
+		add_action( 'wp_enqueue_scripts', [
+			$this->container->make( Early_Access::class ),
+			'maybe_deregister_rsvp_assets'
+		], 100 );
```
* :no_entry_sign: Each array item in a multi-line array declaration must end in a comma
* :no_entry_sign: Comma required after last value in array declaration

<sub>hash: 023e632b8c5d39513302fda681c63164</sub>

tr1b0t commented on src/Tribe/RSVP/Service_Provider.php:
```diff
@@ -0,0 +1,44 @@
+			$this->container->make( Early_Access::class ),
+			'maybe_deregister_rsvp_assets'
+		], 100 );
+
```
* :no_entry_sign: Only one argument is allowed per line in a multi-line function call
* :no_entry_sign: Closing parenthesis of a multi-line function call must be on a line by itself

<sub>hash: 387d3552f0463074cbf1527bb43b7139</sub>

tr1b0t commented on src/Tribe/RSVP/Service_Provider.php:
```diff
@@ -0,0 +1,44 @@
+		], 100 );
+
+		add_filter( 'tribe_events_tickets_template_tickets/rsvp.php', [
+			$this->container->make( Early_Access::class ),
```
* :no_entry_sign: Opening parenthesis of a multi-line function call must be the last content on the line
* :no_entry_sign: Only one argument is allowed per line in a multi-line function call

<sub>hash: 05ce21491297726b760a1db5d8e2b35c</sub>

tr1b0t commented on src/Tribe/RSVP/Service_Provider.php:
```diff
@@ -0,0 +1,44 @@
+		add_filter( 'tribe_events_tickets_template_tickets/rsvp.php', [
+			$this->container->make( Early_Access::class ),
+			'maybe_override_template'
+		] );
```
* :no_entry_sign: Each array item in a multi-line array declaration must end in a comma
* :no_entry_sign: Comma required after last value in array declaration

<sub>hash: fa7c10988701439033e84d67c7bff290</sub>

tr1b0t commented on src/Tribe/RSVP/Service_Provider.php:
```diff
@@ -0,0 +1,44 @@
+			$this->container->make( Early_Access::class ),
+			'maybe_override_template'
+		] );
+	}
```
:no_entry_sign: Closing parenthesis of a multi-line function call must be on a line by itself

<sub>hash: d2c65273213ad77a40cb87e397e28d59</sub>

tr1b0t commented on src/Tribe/RSVP/Service_Provider.php:
```diff
@@ -0,0 +1,44 @@
+		] );
+	}
+}
\ No newline at end of file
```
:no_entry_sign: File must end with a newline character

<sub>hash: 56ef6ec5c4654aa4bf791348977b533b</sub>

juanfra commented:
Great work on this one! I have a couple thoughts though.

- I think conceptually, having this as a `Early_Access` means that it purpose is only valid until we release. Meaning that if we want to keep it as a toggle, we'll have to rewrite it to make sense. Or at least re-wording. I'd go with a toggle instead. Like `Activate new RSVP block` or something.
- I'd have cascade checks, and effectively use a **constant**, and probably a filter. We're relying on an option check, and that means we need a plugin to set the option. I'd include a template tag for the check, like we did in views V2: https://github.com/moderntribe/the-events-calendar/blob/bbd78d58e58bf3e50ed5404e173c068f3314a514/src/functions/views/provider.php#L14
- I get the point of having these checks and assets modifications particularly here, it's testable. But 
	- Eventually we'll have to do this in the plugin, where the files are included. Meaning that we'll have to move the functionality where we are not trying to touch now (Assets, or the template inclusions).
	- By hooking the template with the `maybe_`, are we doing it so we replace the template for the classic and block? We need to be sure that now that we're unifying the template, it receives the same information. That's why I feel a simple IF check before the template inclusion would be more straightforward. And also, because at some point we'll need to do it there when we decide to release this.

Luc45 commented:
@juanfra I've refactored it to use a constant instead of an option, and added a filter for testability, thanks.

If we don't plan to keep the old RSVP hanging around, perhaps we can just override it, then? And you'd work directly on `rsvp.php`, `rsvp.css` and `rsvp.js`

Luc45 commented:
I've also broken down `Early_Access` into smaller classes. I felt the need to do so when I went to work on my other ticket to add an "Update Notice".

After adding it there, it felt like having too much stuff on that class. Out of curiosity, this aims to comply with two SOLID principles: **S**ingle Responsibility Principle, and the **O**pen-Closed Principle.

bordoni commented on src/Tribe/RSVP/Early_Access/Early_Access.php:
```diff
@@ -0,0 +1,29 @@
+<?php
+
+namespace Tribe\Tickets\RSVP\Early_Access;
+
+
+/**
+ * Class Early_Access
+ *
+ * Handles Early Access for the new RSVP template.
+ *
+ * @since TBD
+ *
+ * @package Tribe\Tickets\RSVP\Early_Access
+ */
+class Early_Access {
+
+	/**
+	 * Returns whether we are using RSVP Early Access or not.
+	 *
+	 * @since TBD
+	 *
+	 * @return bool Whether it's early access or not.
+	 */
+	public function is_rsvp_early_access() {
+		$is_early_access = defined( 'TRIBE_TICKETS_RSVP_EARLY_ACCESS' ) && tribe_is_truthy( TRIBE_TICKETS_RSVP_EARLY_ACCESS );
+
+		return (bool) apply_filters( 'tribe_tickets_is_rsvp_early_access', $is_early_access );
+	}
+}
```
I would rather see this live in a template tag that gets included manually inside of the service provider, following what we did with View V2;

https://github.com/moderntribe/the-events-calendar/blob/master/src/functions/views/provider.php#L14-L35

bordoni commented on src/Tribe/RSVP/Early_Access/Assets.php:
```diff
@@ -0,0 +1,54 @@
+	 * @see \Tribe\Tickets\RSVP\Service_Provider::register_early_access
+	 */
+	public function deregister_rsvp_assets() {
+		/* @see \Tribe__Tickets__RSVP::register_resources */
+		wp_deregister_style( 'event-tickets-rsvp' );
+		wp_deregister_script( 'event-tickets-rsvp' );
+
+		/* @see \Tribe__Tickets__Assets::enqueue_scripts */
+		wp_deregister_style( 'event-tickets-tickets-rsvp-css' );
+		wp_deregister_script( 'event-tickets-tickets-rsvp-js' );
```
If this is using `tribe_asset` we should use `Tribe__Assets::remove()`

bordoni commented on src/Tribe/RSVP/Service_Provider.php:
```diff
@@ -0,0 +1,53 @@
+	 *
+	 * @since TBD
+	 */
+	public function register() {
+		$this->register_early_access();
+	}
```
Check the pattern we used on Views V2 with Hooks, Assets, and registration of the Service Provider.
https://github.com/moderntribe/the-events-calendar/blob/master/src/Tribe/Views/V2/Service_Provider.php

bordoni commented on src/Tribe/RSVP/Early_Access/Template.php:
```diff
@@ -0,0 +1,34 @@
+<?php
+
+namespace Tribe\Tickets\RSVP\Early_Access;
```
We should avoid calling it Early Access because that will be the final name.
Maybe call it `V2` or `Unified`, not sure. The Green Devs should agree on a final name since we want to avoid deprecations.

sc0ttkclark commented:
Going to rework the changes here off of G20.06 branch and implement code review changes requested

sc0ttkclark commented:
Replaced with #1683
